MeesGame/AI/FloodFill.cs

```diff
@@ -0,0 +1,176 @@
+            {
+                player.NextAIAction = PlayerAction.NONE;
+            }
+        }
+
+        /// <summary>
+        /// The main method of the FloodFill AI.
+        /// </summary>
+        /// <param name="p"> Point to start/continue the flooding from. </param>
+        private void Flood(Point p)
+        {
+            // If we have already visited this tile, return.
+            if (visited[p.X, p.Y])
+                return;
+
+            // The current tile has been visited.
+            visited[p.X, p.Y] = true;
+
+            // GuardTile is subclass of HoleTile so this will return for both.
+            // Can't use the same for walls because DoorTile is a subclass of WallTile.
+            if (tileField.GetTile(p.X, p.Y) is HoleTile || tileField.GetType(p.X, p.Y) == TileType.Wall)
+                return;
+
+            // Add the current tile to the route.
+            currentRoute.Push(p);
+
+            // Check if the current tile is a keytile,
+            // and whether we have already picked up a key of this color.
+            if (tileField.GetTile(p.X, p.Y).TileType == TileType.Key &&
+                (pickedUpKeys & keys[(tileField.GetTile(p.X, p.Y) as KeyTile).KeyColor]) == 0)
+            {
+                // Add the key to the picked up keys.
+                pickedUpKeys |= keys[(tileField.GetTile(p.X, p.Y) as KeyTile).KeyColor];
+
+                // We need to flood everything again because there could be doors that are open now.
+                UnvisitEverything();
+
+                // The tile we are currently on is still visited.
+                visited[p.X, p.Y] = true;
+            }
+
+            foreach (PlayerAction action in Enum.GetValues(typeof(PlayerAction)))
+            {
+                if (action == PlayerAction.SPECIAL && tileField.GetTile(p.X, p.Y).TileType == TileType.Portal)
+                    Flood((tileField.GetTile(p) as PortalTile).Destination);
+                // Using else if (action != PlayerAction.NONE) cause a bug somehow.
+                else if (action == PlayerAction.NORTH || action == PlayerAction.EAST || action == PlayerAction.SOUTH || action == PlayerAction.WEST)
+                {
+                    // Obstacles are checked earlier in Flood() so we only need to check whether the new point is in the tilefield.
+                    if (!tileField.OutOfTileField(p + action.ToDirection().ToPoint()))
+                    {
+                        Flood(p + action.ToDirection().ToPoint());
+                    }
+                }
+            }
+
+            if (tileField.GetTile(p.X, p.Y).TileType == TileType.End)
+            {
+                // If there is no best route yet.
+                if (bestRoute.Count == 0)
+                    bestRoute = new Stack<Point>(new Stack<Point>(currentRoute));
+                // If the current route is better than the curent best route.
+                else if (bestRoute.Count > currentRoute.Count)
```

Comment:
Alright, how do I stop all the Flood methods currently running then? If I just do return it'll keep going anyway.